src/main/java/com/coreos/jetcd/internal/impl/ClientImpl.java

```diff
@@ -13,178 +10,152 @@
 import com.coreos.jetcd.Lease;
 import com.coreos.jetcd.Maintenance;
 import com.coreos.jetcd.Watch;
-import com.coreos.jetcd.api.AuthGrpc;
-import com.coreos.jetcd.api.AuthenticateRequest;
-import com.coreos.jetcd.api.AuthenticateResponse;
 import com.coreos.jetcd.exception.AuthFailedException;
 import com.coreos.jetcd.exception.ConnectException;
-import com.coreos.jetcd.internal.Pair;
-import com.google.common.base.Supplier;
-import com.google.common.base.Suppliers;
-import com.google.common.util.concurrent.ListenableFuture;
-import com.google.protobuf.ByteString;
 import io.grpc.ManagedChannel;
 import io.grpc.ManagedChannelBuilder;
 import io.grpc.NameResolver;
 import java.util.Optional;
-import java.util.concurrent.ExecutionException;
-import java.util.concurrent.ExecutorService;
-import java.util.concurrent.Executors;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
 
 /**
  * Etcd Client.
  */
-public class ClientImpl implements Client {
-
-  private final NameResolver.Factory nameResolverFactory;
-  private final Supplier<KV> kvClient;
-  private final Supplier<Auth> authClient;
-  private final Supplier<Maintenance> maintenanceClient;
-  private final Supplier<Cluster> clusterClient;
-  private final Supplier<Lease> leaseClient;
-  private final Supplier<Watch> watchClient;
-  private final ByteString user;
-  private final ByteString pass;
-  private final ManagedChannel channel;
-  private final Optional<String> token;
-
-  // shared executorService
-  ExecutorService executorService = Executors.newCachedThreadPool();
-
-  ExecutorService getExecutorService() {
-    return this.executorService;
-  }
-
-  ManagedChannel getChannel() {
-    return this.channel;
-  }
-
-  Optional<String> getToken() {
-    return this.token;
-  }
-
-  public ClientImpl(ClientBuilder builder) throws ConnectException, AuthFailedException {
-    this(Optional.empty(), builder);
-  }
-
-  public ClientImpl(ManagedChannelBuilder<?> channelBuilder, ClientBuilder clientBuilder)
-      throws ConnectException, AuthFailedException {
-    this(Optional.ofNullable(channelBuilder), clientBuilder);
-  }
-
-  private ClientImpl(Optional<ManagedChannelBuilder<?>> channelBuilderOptional,
-                     ClientBuilder clientBuilder) throws ConnectException, AuthFailedException {
-    if (clientBuilder.getNameResolverFactory() != null) {
-      this.nameResolverFactory = clientBuilder.getNameResolverFactory();
-    } else {
-      //If no nameResolverFactory was set, use SimpleEtcdNameResolver
-      checkNotNull(clientBuilder.getEndpoints(), "endpoints can't be null");
-      this.nameResolverFactory = ClientUtil.simpleNameResolveFactory(clientBuilder.getEndpoints());
+public final class ClientImpl implements Client {
```

Comment:
maybe add java doc to Client interface about potential Auth and Connect exceptions when init `client` and `get*Client`?